libraries/AP_PacketDigitalCAN/AP_PacketDigitalCAN.cpp

```diff
@@ -0,0 +1,609 @@
+        // nothing to do
+        return;
+    }
+    MPPT_devices[index].timestamp_ms = AP_HAL::millis();
+}
+
+void AP_PacketDigitalCAN::send_command(const PacketType type, const uint16_t serialnumber, const float data)
+{
+    AP_HAL::CANFrame txFrame;
+
+    switch (type) {
+    case PacketType::STREAM_FAULT:
+    case PacketType::STREAM_INPUT:
+    case PacketType::STREAM_OUTPUT:
+    case PacketType::FAULT:
+    case PacketType::ACK:
+    case PacketType::NACK:
+    case PacketType::UNKNOWN:
+        // we only receive these
+        return;
+
+    case PacketType::ALGORITHM_GET:
+    case PacketType::VOLTAGE_GET:
+    case PacketType::CVT_GET:
+    case PacketType::PING:
+        txFrame.dlc = 0;
+        break;
+
+    case PacketType::ALGORITHM_SET:
+        txFrame.dlc = 1;
+        txFrame.data[2] = data;
+        break;
+
+    case PacketType::VOLTAGE_SET:
+    case PacketType::CVT_SET:
+        {
+        txFrame.dlc = 2;
+        const uint16_t value = float2fixed(data);
+        txFrame.data[2] = HIGHBYTE(value);
+        txFrame.data[3] = LOWBYTE(value);
+        }
+        break;
+    }
+
+    if (serialnumber == 0) {
+        // send to all
+        txFrame.id = 0x00240000;
+    } else {
+        txFrame.id = 0x00210000 | (uint32_t)serialnumber;
+    }
+
+    txFrame.id |= AP_HAL::CANFrame::FlagEFF;
+
+    txFrame.data[0] = ++sequence_prev;
+    txFrame.data[1] = (uint8_t)type;
+    txFrame.dlc += 2;
+
+    //gcs().send_text(MAV_SEVERITY_INFO, "PDCAN: %u Tx Seq %u 0x%2X", (unsigned)serialnumber, (unsigned)txFrame.data[0], (unsigned)txFrame.data[1]);
+
+    if (write_frame(txFrame, 50000)) {
+        // keep track of what we sent last in case we get an ACK/NACK
+        packet_sent_prev = type;
+    } else {
+        //gcs().send_text(MAV_SEVERITY_DEBUG, "PDCAN: %u Tx Seq %u 0x%2X failed", (unsigned)serialnumber, (unsigned)txFrame.data[0], (unsigned)txFrame.data[1]);
+    }
+}
+
+// get MPPT_device index by serial number.
+// if serial number found in MPP_devices list, return the 0 indexed value
+// else return UINT8_MAX
+uint8_t AP_PacketDigitalCAN::get_device_index(const uint16_t serial_number) const
+{
+    for (uint8_t i=0; i<device_count; i++) {
+        if (MPPT_devices[i].serialnumber == serial_number) {
+            return i;
+        }
+    }
+    return UINT8_MAX;
+}
+
+// write frame on CAN bus, returns true on success
+bool AP_PacketDigitalCAN::write_frame(AP_HAL::CANFrame &out_frame, const uint64_t timeout)
+{
+    if (!_initialized) {
+        debug_can(AP_CANManager::LOG_ERROR, "Driver not initialized for write_frame\n\r");
+        return false;
+    }
+
+    bool read_select = false;
+    bool write_select = true;
+    bool ret;
+    do {
+        ret = _can_iface->select(read_select, write_select, &out_frame, AP_HAL::native_micros64() + timeout);
+        if (!ret || !write_select) {
+            hal.scheduler->delay_microseconds(200);
+        }
+    } while (!ret || !write_select);
```

Comment:
Is it OK to delay and do this infinite loop?  This looks like its on its own thread so probably OK

Reply:
I removed it in favor of following how AP_PiccoloCAN and AP_CANTester did it

Reply:
This has been removed in favor of using the CANDriver class